kauri/kauri.go

```diff
@@ -0,0 +1,320 @@
+// Package kauri contains the implementation of the kauri protocol
+package kauri
+
+import (
+	"context"
+	"encoding/binary"
+	"errors"
+	"math/rand"
+	"reflect"
+	"sort"
+	"time"
+
+	"github.com/relab/gorums"
+	"github.com/relab/hotstuff"
+	"github.com/relab/hotstuff/backend"
+	"github.com/relab/hotstuff/eventloop"
+	"github.com/relab/hotstuff/internal/proto/hotstuffpb"
+	"github.com/relab/hotstuff/internal/proto/kauripb"
+	"github.com/relab/hotstuff/logging"
+	"github.com/relab/hotstuff/modules"
+)
+
+func init() {
+	modules.RegisterModule("kauri", New)
+}
+
+// TreeType determines the type of tree to be built, for future extension
+type TreeType int
+
+// NoFaultTree is a tree without any faulty replicas in the tree, which is currently supported.
+// TreeWithFaults is a tree with faulty replicas, not yet implemented.
+const (
+	NoFaultTree    TreeType = 1
+	TreeWithFaults TreeType = 2
+)
+
+// Kauri structure contains the modules for kauri protocol implementation.
+type Kauri struct {
+	configuration          *backend.Config
+	server                 *backend.Server
+	blockChain             modules.BlockChain
+	crypto                 modules.Crypto
+	eventLoop              *eventloop.EventLoop
+	logger                 logging.Logger
+	opts                   *modules.Options
+	synchronizer           modules.Synchronizer
+	leaderRotation         modules.LeaderRotation
+	tree                   TreeConfiguration
+	initDone               bool
+	aggregatedContribution hotstuff.QuorumSignature
+	blockHash              hotstuff.Hash
+	currentView            hotstuff.View
+	senders                []hotstuff.ID
+	nodes                  map[hotstuff.ID]*kauripb.Node
+	isAggregationSent      bool
+}
+
+// New initializes the kauri structure
+func New() modules.Kauri {
+	return &Kauri{nodes: make(map[hotstuff.ID]*kauripb.Node)}
+}
+
+// InitModule initializes the Handel module.
+func (k *Kauri) InitModule(mods *modules.Core) {
+	mods.Get(
+		&k.configuration,
+		&k.server,
+		&k.blockChain,
+		&k.crypto,
+		&k.eventLoop,
+		&k.logger,
+		&k.opts,
+		&k.leaderRotation,
+		&k.synchronizer,
+	)
+	k.opts.SetShouldUseKauri()
+	k.eventLoop.RegisterObserver(backend.ConnectedEvent{}, func(_ any) {
+		k.postInit()
+	})
+	k.eventLoop.RegisterHandler(ContributionRecvEvent{}, func(event any) {
+		k.OnContributionRecv(event.(ContributionRecvEvent))
+	})
+}
+
+func (k *Kauri) postInit() {
+	k.logger.Info("Handel: Initializing")
+	kauripb.RegisterKauriServer(k.server.GetGorumsServer(), serviceImpl{k})
+	k.initializeConfiguration()
+}
+
+func (k *Kauri) initializeConfiguration() {
+	kauriCfg := kauripb.ConfigurationFromRaw(k.configuration.GetRawConfiguration(), nil)
+	for _, n := range kauriCfg.Nodes() {
```

Comment:
Seems better to move this into the tree constructor like this:
```go
func newFaultFreeTree(cfg kauripb.Configuration, myID hotstuff.ID) *FaultFreeTree {
	nodes := make(map[hotstuff.ID]*kauripb.Node)
	idMappings := make(map[hotstuff.ID]int)
	posToIDMapping := make(map[int]hotstuff.ID)
	for i, n := range cfg.Nodes() {
		nodes[hotstuff.ID(n.ID())] = n
		idMappings[hotstuff.ID(i+1)] = i
		posToIDMapping[i] = hotstuff.ID(i + 1)
	}
	// some code removed for brevity
	return &FaultFreeTree{
		ID:                  myID,
		ConfigurationLength: len(nodes),
		height:              height,
		nodes:               nodes,
		idToPosMapping:      idMappings,
		posToIDMapping:      posToIDMapping,
	}
}
```

Note: I don't think you actually need all three maps, but I didn't check. 
Also, it seems the `hotstuff.ID` generation in the above code could be using only a single one like this:
```go
	id := hotstuff.ID(n.ID())
	nodes[id] = n
	idMappings[id] = i
	posToIDMapping[i] = id
```

Reply:
This said, I think a more idiomatic design would be to pass the `TreeConfiguration` as an argument to the `Karui` module. But I don't know how to do that with the module system.